### PR TITLE
C 03 01 draft

### PR DESCRIPTION
**Pull request by uhasker**

### CODE REVIEW

File: src/SUMMARY.md [hunk 1]
```diff
@@ -13,9 +13,13 @@
   - [Functional thinking](chapter1/06-functional-thinking.md)
 
 - [The client](chapter2/README.md)
+
```
Comment (nyarthan) [inline src/SUMMARY.md:16]: why is here a line?

File: src/SUMMARY.md [hunk 2]
```diff
@@ -13,9 +13,13 @@
...
   - [HTML and the DOM](chapter2/01-html-and-the-dom.md)
   - [First steps in React](chapter2/02-first-steps-in-react.md)
   - [Introducing JSX](chapter2/03-introducing-jsx.md)
   - [React components](chapter2/04-react-components.md)
   - [Project structure](chapter2/05-project-structure.md)
   - [React state](chapter2/06-react-state.md)
+
+- [The server](chapter3/README.md)
```
Comment (nyarthan) [inline src/SUMMARY.md:24]: but not here? hmm

File: src/chapter3/01-a-http-primer.md [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+
+### HTTP and Express.js
+
+When writing a web application, the most common way to get data from a server to a client is via the HTTP procotol (usually augmented by SSL resulting in the HTTPS procotol).
+HTTP is a **request/response protocol**, i.e. HTTP clients send requests to an HTTP server and receive a response in return.
+You have in fact already used the HTTP protocol.
+Every time you browse the internet, HTTP requests are sent under the hood.
+
+We could use the built-in HTTP module of Node.js, but this is not terribly convenient.
+Instead we will use an extremely popular framework called **Express.js**.
```
Comment (nyarthan) [inline src/chapter3/01-a-http-primer.md:11]: its just called Express

File: src/chapter3/01-a-http-primer.md [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+npm init -y
+```
+
+Install Express.js:
+
+```shell
+npm install express
+```
+
+Now create a file named server.js inside the api directory.
```
Comment (nyarthan) [inline src/chapter3/01-a-http-primer.md:28]: The express guide uses app.js. Maybe we should stick to that convention

File: src/chapter3/01-a-http-primer.md [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+A port is a communication endpoint on that host.
+The concept of ports allows you to run multiple network applications on the same device.
+For example, your awesome website might run on port 80 and a service allowing you to remotely login onto your machine might run on port 22.
+
+This is how you can create an Express application and make it listen on port 3000:
+
+```javascript
+const express = require('express');
+const app = express();
+const port = 3000;
```
Comment (nyarthan) [inline src/chapter3/01-a-http-primer.md:42]: constantts are SCREAMING_SNAKE_CASE

File: src/chapter3/01-a-http-primer.md [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+For didactic reasons we will include the port:
+
+```
+https://www.google.com:443/search?q=mern+stack
+```
+
+Generally speaking, a typical HTTP URL has the following form:
+
+```
+scheme://host:port/path?key1=value1&key2=value2#test
```
Comment (nyarthan) [inline src/chapter3/01-a-http-primer.md:97]: "#test" should be "#hash" or "#fragment"

File: src/chapter3/01-a-http-primer.md [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+```
+
+> Note the word _typical_ here.
+> HTTP URLs (and especially more general URLs) can become _much more_ complicated, but we will not cover all the little details in this book.
+> In fact the complexity of URLs is often a source of subtle bugs and browser crashes.
+> For example - at some point Android Chrome would [crash](https://news.ycombinator.com/item?id=28639708) when trying to open `http://../foo`.
+
+The **scheme** specifies the scheme used and usually indicates the protocol.
```
Comment (nyarthan) [inline src/chapter3/01-a-http-primer.md:105]: apparently the scheme specifies the scheme
maybe explain what a scheme is?

File: src/chapter3/01-a-http-primer.md [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+The path can be followed by a **query**.
+The query begins with a question mark and is followed by key-value pairs.
+In the Google URL this is "?q=mern+stack".
+Here the query provides information about your search.
+If there are multiple key-value pairs, they are separated by ampersands ("&").
+For example the query could be "?key1=value1&key2=value2".
+
+The query can be followed by a **fragment**.
+This is used for navigation by the client and is not sent to server.
+We will return to fragment a bit later in the book.
```
Comment (nyarthan) [inline src/chapter3/01-a-http-primer.md:133]: will we? unless we build a blog or sth I don't really se a use for them

File: src/chapter3/01-a-http-primer.md (hunk 3 above)
Comment (uhasker) [reply src/chapter3/01-a-http-primer.md:11]: its called both, but the official name is Express, yes

File: src/chapter3/01-a-http-primer.md (hunk 8 above)
Comment (uhasker) [reply src/chapter3/01-a-http-primer.md:133]: we probably will, but this sentence is a bit useless anyway, so should strike it

File: src/chapter3/01-a-http-primer.md [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+This application is not very useful, so let's add a **route** to it.
+Add this above the call to `listen`:
+
+```javascript
+app.get('/', (req, res) => {
+	res.send('Hello, world!');
+}
+```
+
+If you open a browser and navigate to <http://localhost:3000> you will see
```
[the remark this answers is not in the document]
Comment (uhasker) [reply src/chapter3/01-a-http-primer.md:58]: fixed

File: src/chapter3/01-a-http-primer.md [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+  console.log(`Server listening on port ${PORT}...`);
+});
+```
+
+This application is not very useful, so let's add a **route** to it.
+Add this above the call to `listen`:
+
+```javascript
+app.get('/', (req, res) => {
+	res.send('Hello, world!');
```
[the remark this answers is not in the document]
Comment (uhasker) [reply src/chapter3/01-a-http-primer.md:54]: fixed

File: src/chapter3/01-a-http-primer.md [hunk 11]
```diff
@@ -0,0 +1,225 @@
+## A HTTP primer
```
[the remark this answers is not in the document]
Comment (uhasker) [reply src/chapter3/01-a-http-primer.md:1]: fixed

**Pull request merged**